### ISSUE TITLE
maven: running before immutables generates source code

### ISSUE THREAD

**Issue by yatesco** [typetools/checker-framework#5143]
Hi - this is more a maven question, but googling has failed me. How do I ensure that the `checkerframework` runs *after* other things in maven that generate source code?

I am using immutables.org which generates source code into `target\generated-sources`. The generated source code is referenced in the test code. However, after adding in the three maven profiles the immutables logic is no longer run which causes the code not to compile.

If I comment out the three build profiles then `mvn clean package` works. If I leave the three build profiles in then `mvn clean package` fails because it can't find the generated source code and `target\generated-sources` is empty.

I've tried adding the `org.immutables` dependency in the checkerframework's profile's `annotationProcessorPaths` but that still didn't generate the source code.

Do you have any pointers I could chase here? Thanks!

**Comment by yatesco** [comment]
I think I've fixed this by adding the `org.immutables.value.internal.$processor$.$Processor` to the list of `annotationProcessors` in the `checkerframework` build profile:
```xml
<profile>
            <id>checkerframework</id>
            <activation>
                <jdk>[1.8,)</jdk>
            </activation>
            <build>
                <plugins>
                    <plugin>
                        <groupId>org.apache.maven.plugins</groupId>
                        <artifactId>maven-compiler-plugin</artifactId>
                        <configuration>
                            <fork>true</fork> <!-- Must fork or else JVM arguments are ignored. -->
                            <compilerArguments>
                                <Xmaxerrs>10000</Xmaxerrs>
                                <Xmaxwarns>10000</Xmaxwarns>
                            </compilerArguments>
                            <annotationProcessorPaths>
                                <path>
                                    <groupId>org.checkerframework</groupId>
                                    <artifactId>checker</artifactId>
                                    <version>3.22.0</version>
                                </path>
                            </annotationProcessorPaths>
                            <annotationProcessors>
                                <!-- ***** THIS IS THE MAGIC ****** -->
                                <annotationProcessor>org.immutables.value.internal.$processor$.$Processor
                                <!-- Add all the checkers you want to enable here -->
                                </annotationProcessor>
                                <annotationProcessor>org.checkerframework.checker.nullness.NullnessChecker
                                </annotationProcessor>
                            </annotationProcessors>
                            <compilerArgs>
                                <!-- <arg>-Awarns</arg> --> <!-- -Awarns turns type-checking errors into warnings. -->
                            </compilerArgs>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
            <dependencies>
                <dependency>
                    <groupId>org.checkerframework</groupId>
                    <artifactId>checker</artifactId>
                    <version>3.22.0</version>
                </dependency>
            </dependencies>
        </profile>
```

I'm at the "throw random XML lines until the errors go away" so I'll leave this open in the hope somebody confirms/upgrades these instructions.

**Comment by msridhar** [comment]
@yatesco I think the issue here is likely that the immutables processor normally runs using "auto-discovery", where just placing it in the processor path causes it to run by default.  Auto-discovery gets disabled if you add _any_ explicit `-processor` arguments (or `<annotationProcessor>` options in Maven) in your config.  This can be fixed either by adding explicit arguments for every annotation processor you want to run (I think this is your current fix), or by running the Checker Framework using auto-discovery, as documented here:

https://checkerframework.org/manual/#checker-auto-discovery

Let us know if that makes sense / solves your problem.

**Comment by yatesco** [comment]
Thanks @msridhar - yes, this worked - thank you. Closing.

**Comment by hubchenfox** [comment]
I looked at your link.
As far as I understood it, it was about two different plugins both being executed in the maven lifecycle compile, but one of them was no longer executed.
Your solution from the link:
If you run "mvn clean generate-sources", the autogenerated classes of both plugins are "missing", aren't they?
I have now got QueryDSL to generate the classes in the compile lifecylce (annotationProcessor) only but this has other disadvantages and that is why I am not so happy with it.
